sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/Rule.scala

```diff
@@ -33,4 +33,31 @@ abstract class Rule[TreeType <: TreeNode[_]] extends SQLConfHelper with Logging
   }
 
   def apply(plan: TreeType): TreeType
+
+  def apply(plan: TreeType, ruleContext: Option[RuleContextBase]): TreeType = apply(plan)
+}
+
+trait RuleContextBase {
+  def isSubquery: Boolean
+
+  private[sql] def withSubquery(isSubquery: Boolean): RuleContextBase
```

Comment:
shall we make it public as well? Third-party rules may need to process subqueries and propagate rule context as well .